### ISSUE TITLE
Deprecate SHA-1 features

### ISSUE THREAD

**Issue by briansmith** [briansmith/ring#754]
1. Every SHA-1 feature should be marked @deprecated.
2. Each SHA-1 feature (RSA-SHA1 signature verification, `ring::digest::SHA1`, HMAC SHA-1, PBKDF2 with SHA-1, etc.) should be disabled by default and enabled with a separate feature.

These two steps should be done in separate PRs, maybe even multiple PRs per step.

#1 is complicated by the fact that some of these things are defined indirectly via macros. I'm not sure yet how to change the macros so that some invocations of them result in a `#[deprecated]` item and the others don't.

#2 is complicated by the fact that we don't have separate `HMAC_SHA1` or `PBKDF2_SHA1` algorithms anymore; instead HMAC and PBKDF2 are parameterized by a digest algorithm, and all digest algorithms are treated equally. OTOH HMAC-SHA1 and PBKDF2-SHA1 are lower-risk so it is OK to start off with making the off-by-default features for `ring::digest::SHA1` and `RSA-SHA1` and punt on doing similar for PBKDF2 and HMAC for a while.

**Comment by pietro** [comment]
I'll try and take a stab at this. For the macros I'll try adding the ability to pass metaitems for the attributes with as little code duplication as possible.

**Comment by briansmith** [comment]
In the current master, I've renamed `SHA1` to `SHA1_FOR_LEGACY_USE_ONLY` and did so throughout *ring* except for PBKDF2.

I had somebody ask me about this because they were concerned we were going to remove the SHA-1 support from the library if we're marking it `#[deprecated]`. I think it's unlikely we'll be able to remove SHA-1 in the next few years and I think marking it `#[deprecated]` will just get people overly worried.

In the future, we may put SHA1 behind a Cargo feature flag. However, that would be predicated on us having CI support for testing the default configuration and the SHA-1 enabled configuration separately. CI is currently a disaster in terms of how long it takes to finish. Also, before we start doing this, we should probably start testing the various `--no-default-features` configurations of *ring* in CI first. So I don't think this will change soon.

I think the new naming will go a long way towards the goal here with minimal confusion so I'm going to close this. @pietro, sorry I wasted your time with this.
